asset/encoding.go

```diff
@@ -664,3 +671,131 @@ func LeafDecoder(r io.Reader, val any, buf *[8]byte, l uint64) error {
 	}
 	return tlv.NewTypeForEncodingErr(val, "Asset")
 }
+
+func EncodeTapBranchPreimage(preimage TapBranchPreimage) []byte {
+	var encodedBranch bytes.Buffer
+	_, _ = encodedBranch.Write(preimage.left[:])
+	_, _ = encodedBranch.Write(preimage.right[:])
+
+	return encodedBranch.Bytes()
+}
+
+func DecodeTapBranchPreimage(branchData []byte) (*TapBranchPreimage, error) {
+	// Given data cannot be a valid encoded TapBranchPreimage.
+	if len(branchData) != TapBranchPreimageLen {
+		return nil, fmt.Errorf("invalid tapscript preimage length")
+	}
+
+	var leftHash, rightHash [chainhash.HashSize]byte
+	copy(leftHash[:], branchData[:chainhash.HashSize])
+	copy(rightHash[:], branchData[chainhash.HashSize:])
+
+	return &TapBranchPreimage{
+		left:  leftHash,
+		right: rightHash,
+	}, nil
+}
+
+// The following TapLeaf {en,de}coders are a duplicate of those used in
+// btcwallet. Specifically, the inner loop logic for handling []TapLeaf.
+// https://github.com/btcsuite/btcwallet/blob/master/waddrmgr/tlv.go#L160
+// This duplication is needed until we update btcwallet to export these methods.
+
+// EncodeTapLeaf encodes a TapLeaf into a byte slice containing a TapLeaf TLV
+// record, prefixed with a varint indicating the length of the record.
+func EncodeTapLeaf(leaf *txscript.TapLeaf) ([]byte, error) {
+	if leaf == nil {
+		return nil, fmt.Errorf("cannot encode nil tapleaf")
+	}
+
+	leafVersion := uint8(leaf.LeafVersion)
+	tlvRecords := []tlv.Record{
+		tlv.MakePrimitiveRecord(
+			typeTapLeafVersion, &leafVersion,
+		),
+	}
+
+	if len(leaf.Script) > 0 {
```

Comment:
Why would the version be specified, but not also the script?

Reply:
Not sure, but I think you could call `txscript.NewTapLead` with an empty script but not an empty version.